src/APIBasedAuth.ts

```diff
@@ -0,0 +1,192 @@
+import axios, {
+  AxiosError,
+  AxiosInstance,
+  AxiosRequestConfig,
+  AxiosResponse
+} from 'axios';
+
+/* istanbul ignore next */
+const formatAxiosError = <ErrorResponse = any>(
+  error: AxiosError<ErrorResponse> | unknown
+) => {
+  if (!axios.isAxiosError(error)) {
+    throw new AxiosError(String(error));
+  }
+  throw error.response?.data;
+};
+
+class APIBasedAuth {
+  private client: AxiosInstance;
+  private clientOptions: APIBasedAuth.Config;
+  private session?: APIBasedAuth.Session;
+  private tokenData?: APIBasedAuth.TokenData;
+
+  constructor(options: APIBasedAuth.Config) {
+    const required = ['baseURL', 'clientId'];
+    for (const param of required) {
+      // eslint-disable-next-line security/detect-object-injection
+      if (!options[param]) {
+        throw new Error(`APIBasedAuth ctor param ${param} is required`);
+      }
+    }
+
+    this.client = axios.create({
+      baseURL: options.baseURL,
+      withCredentials: options.withCredentials
+    });
```

Comment:
We seem to be ignoring most of the `options` inputs here. I think we should either:
- not accept the general axios config in the constructor, or
- spread the config into this object

Reply:
If we go for Option 1, we get the benefit of not exposing any `axios`-specific API, which would be cool.

Reply:
within TS I've made the config pick only baseURL and withCredentials from axios